### PR TITLE
try to refactor container/bytejson

### PR DESCRIPTION
**Pull request by forsaken628**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [x] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #15725

## What this PR does / why we need it:
Improve performance

### CODE REVIEW

**Comment by matrix-meow** [comment]
@forsaken628 Thanks for your contributions!

**Here are review comments for file go.mod:**
### Pull Request Review:

#### Title: 
The title of the pull request is vague and does not provide much information about the specific changes being made. It would be beneficial to have a more descriptive title that clearly indicates the purpose of the refactoring.

#### Body:
The body of the pull request is concise and mentions that the goal is to improve performance. It would be helpful to include more details about the performance improvements being targeted to provide better context for the changes.

#### Changes in go.mod:
1. **Addition of `github.com/json-iterator/go v1.1.12`:**
   - **Issue:** The addition of a new dependency (`github.com/json-iterator/go`) without a clear explanation in the pull request body raises concerns about the necessity and impact of this change.
   - **Security Concern:** Introducing a new dependency can potentially introduce security vulnerabilities if the dependency is not thoroughly vetted.
   - **Suggestion:** Provide a detailed explanation in the pull request body regarding why this specific dependency was added, what performance improvements it brings, and how it aligns with the project's requirements. Additionally, ensure that the added dependency is secure and actively maintained.

2. **Removal of `github.com/json-iterator/go v1.1.12` from the `indirect` dependencies:**
   - **Issue:** The removal of the dependency from the `indirect` dependencies without a clear reason can lead to confusion and may impact the project's functionality.
   - **Suggestion:** If the dependency is no longer needed as an `indirect` dependency, provide a brief explanation in the pull request body clarifying the rationale behind this change.

#### Suggestions for Optimization:
1. **Detailed Explanation:**
   - Provide a comprehensive explanation in the pull request body detailing the reasons for each change made in the `go.mod` file. This will help reviewers and maintainers understand the purpose of the modifications.

2. **Security Review:**
   - Conduct a thorough security review of the newly added dependency (`github.com/json-iterator/go v1.1.12`) to ensure that it does not introduce any security vulnerabilities to the project.

3. **Consistency in Dependency Management:**
   - Ensure consistency in managing dependencies by clearly stating the reasons for adding or removing dependencies in the project.

4. **Testing:**
   - Consider adding tests to validate the performance improvements achieved through the refactoring to ensure that the changes have the desired impact.

By addressing the mentioned issues and following the optimization suggestions, the pull request can be enhanced to provide a clearer understanding of the changes made and their impact on the project.

**Here are review comments for file pkg/container/bytejson/bytejson_test.go:**
### Pull Request Review:

#### Title:
The title of the pull request is vague and does not provide clear information about the changes being made. It should be more descriptive to give a better understanding of the purpose of the changes.

#### Body:
The body of the pull request is brief and lacks detailed information about the improvements being made. It would be beneficial to include more specific details about the performance enhancements being implemented.

#### Changes in `pkg/container/bytejson/bytejson_test.go`:
1. **Addition of `encoding/json` Import:**
   - The addition of the `encoding/json` import seems unnecessary in the context of the existing test file. If it is not being used in the test cases or related functions, it should be removed to avoid unnecessary imports.

2. **Benchmark Functions:**
   - The benchmark functions `BenchmarkParseJsonByteFromString` and `BenchmarkParseJsonByteFromString2` are added to measure the performance of parsing JSON from a string. These benchmarks are useful for performance optimization.
   - Ensure that the benchmark functions are accurately measuring the performance improvements intended and that the results are reliable.
   - Consider adding comments to explain the purpose of each benchmark function for better clarity.

3. **Fuzz Function:**
   - The `FuzzParseJsonByteFromString` function is added for fuzz testing JSON parsing from a string.
   - It is important to ensure that the fuzz testing covers a wide range of inputs to validate the robustness of the JSON parsing functionality.
   - Verify that the fuzz testing function is effectively testing edge cases and handling invalid inputs appropriately.

#### Suggestions for Optimization:
1. **Title and Body:**
   - Update the title to reflect the specific refactoring or performance improvement being made.
   - Provide a more detailed description in the body explaining the performance enhancements achieved through the changes.

2. **Unused Imports:**
   - Remove the `encoding/json` import if it is not utilized in the test file to keep the code clean and avoid unnecessary dependencies.

3. **Benchmark Functions:**
   - Ensure that the benchmark functions are accurately measuring the performance improvements and are well-documented for future reference.
   - Consider adding comments to describe the expected behavior and purpose of each benchmark.

4. **Fuzz Testing:**
   - Verify that the fuzz testing function covers a diverse set of inputs to validate the JSON parsing functionality thoroughly.
   - Ensure that the fuzz testing function handles invalid inputs gracefully and provides meaningful test results.

By addressing the mentioned points and optimizing the changes made in the pull request, the codebase can be improved in terms of clarity, performance, and robustness.

**Here are review comments for file pkg/container/bytejson/utils.go:**
### Pull Request Review:

#### Title:
The title of the pull request is vague and does not provide much information about the specific changes being made. It would be beneficial to have a more descriptive title that clearly indicates the purpose of the refactoring being done.

#### Body:
The body of the pull request lacks detailed information about the changes being made. While it mentions improving performance, it would be helpful to include specifics on how the performance is being improved and what aspects of the code are being optimized.

#### Changes:
1. **Security Issue:**
   - The change made in the `toString` function is potentially introducing a security vulnerability. Using `json.Marshal` on user-provided input (in this case, `string(data)`) can lead to JSON injection attacks. It's important to sanitize and validate the input data before encoding it as JSON to prevent malicious input from causing unexpected behavior.

2. **Code Optimization:**
   - The change from `strconv.AppendQuote` to `json.Marshal` in the `toString` function may not be the most efficient way to achieve the desired result. `json.Marshal` is typically used for encoding data structures to JSON, not for quoting strings. Consider revisiting this change and using a more appropriate method for quoting the string data.

3. **Error Handling:**
   - The `_` in `bs, _ := json.Marshal(string(data))` ignores any error returned by `json.Marshal`. It's important to handle errors properly to ensure robust error management in the code. Consider checking and handling any errors returned by `json.Marshal`.

#### Suggestions for Improvement:
1. **Security Enhancement:**
   - Implement input validation and sanitization before encoding data as JSON to prevent JSON injection attacks. Consider using a safer method for quoting strings, such as `strconv.Quote` or custom sanitization functions.

2. **Code Optimization:**
   - Revisit the use of `json.Marshal` in the `toString` function and ensure it aligns with the intended functionality. If quoting strings is the goal, explore alternative methods that are more suitable for this purpose.

3. **Error Handling Improvement:**
   - Update the code to handle errors returned by `json.Marshal` appropriately. This ensures that any potential issues during JSON encoding are captured and managed effectively.

By addressing the security concerns, optimizing the code changes, and improving error handling, the pull request can enhance the overall quality and security of the codebase.

File: pkg/container/bytejson/bytejson_test.go [hunk 1]
```diff
@@ -514,3 +517,63 @@ func TestByteJson_Unquote(t *testing.T) {
...
+	// pkg: github.com/matrixorigin/matrixone/pkg/container/bytejson
+	// cpu: Intel(R) Core(TM) i3-9100F CPU @ 3.60GHz
+	// === RUN   BenchmarkParseJsonByteFromString2
+	// BenchmarkParseJsonByteFromString2
+	// BenchmarkParseJsonByteFromString2-4       243330              4863 ns/op            1416 B/op         37 allocs/op
+}
+
+func FuzzParseJsonByteFromString(f *testing.F) {
+	f.Add(`{"a":{"b":{"c":{"d":[null,false,true,123,"abc",[1,2,3],{"a":1,"b":2,"c":3,"d":4,"e":5},123.456]}}}}`)
+	f.Add("0000")
```
Comment (forsaken628) [inline pkg/container/bytejson/bytejson_test.go:551]: too many corner case

File: pkg/container/bytejson/bytejson.go [hunk 2]
```diff
@@ -673,3 +680,293 @@ func ParseJsonByteFromString(s string) ([]byte, error) {
...
+}
+
+func (g *group) reset() {
+	g.obj = false
+	g.keys = g.keys[:0]
+	g.values = g.values[:0]
+}
+
+func (p *parser) parseNumber(in json.Number) (TpCode, []byte, error) {
+	if !json.Valid([]byte(in)) {
```
Comment (forsaken628) [inline pkg/container/bytejson/bytejson.go:733]: may remove this when finded a way to fix corner issues

**Comment by forsaken628** [comment]
tried https://pkg.go.dev/encoding/json#Decoder.Token , too slow

**Comment by forsaken628** [comment]
Solution2
https://github.com/forsaken628/matrixone/blob/json2/pkg/container/bytejson/decode.go

copy from https://github.com/protocolbuffers/protobuf-go/blob/v1.33.0/internal/encoding/json/decode.go

more safety than json-iterator，but slower than json-iterator

**Comment by forsaken628** [comment]
env:
goos: linux
goarch: amd64
pkg: github.com/matrixorigin/matrixone/pkg/container/bytejson
cpu: Intel(R) Core(TM) i3-9100F CPU @ 3.60GHz

std go parser with map ast:
BenchmarkParseJsonByteFromString-4        190492              6829 ns/op            4616 B/op         60 allocs/op

json-iterator parser with slice ast:
BenchmarkParseJsonByteFromString2-4       243330              4863 ns/op            1416 B/op         37 allocs/op

pb parser with slice ast:
BenchmarkParseJsonByteFromString2-4       213036              5601 ns/op            1304 B/op         31 allocs/op

**Review by m-schen — APPROVED** [review]

**Review by heni02 — APPROVED** [review]

**Review by zhangxu19830126 — APPROVED** [review]

**Review by YANGGMM — APPROVED** [review]

**Review by XuPeng-SH — APPROVED** [review]

**Pull request merged**